[Scheduler] Proposal - initial structure for Symfony Scheduler documentation (#19244)
symfony/symfony-docs: scheduler.rst

Scheduler
=========

.. versionadded:: 6.3

    The Scheduler component was introduced in Symfony 6.3

The Symfony Scheduler is a powerful and flexible component designed to manage tasks scheduling within your PHP application.

This document focuses on using the Scheduler component in the context of a full stack Symfony application.

Installation
------------

In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
install the scheduler component:

.. code-block:: terminal

    $ composer require symfony/scheduler

> Introduction to the case:

Reviewer: ```suggestion
Introduction to the case
```

-------------------------

> Embarking on a task is one thing, but often, the need to repeat that task looms large. While one could resort to issuing commands and scheduling them with cron jobs, this approach involves external tools and additional configuration. The Symfony Scheduler component emerges as a solution, allowing us to retain control, configuration, and maintenance of task scheduling within our PHP application.

Reviewer: As far as I am aware, we never use "us". You may replace by "allowing you to retain control". Same for "our", to be replace by "your"

At its core, the principle is straightforward: a task, considered as a message needs to be managed by a service, and this cycle must be repeated.

> Does this sound familiar? Think :doc:`Symfony Messenger docs </components/messenger>`.

Reviewer: I don't know if we use that much of question-like sentences. I guess we could go for something like "This may sound familiar if you're already used to Symfony Messenger" for example

> But while the system of Symfony Messenger proves very useful in various scenarios, there are instances where its capabilities fall short, particularly when dealing with repetitive tasks at regular intervals.

Reviewer: ```suggestion
While the system of Messenger proves very useful in various scenarios, there are instances where its capabilities fall short, particularly when dealing with repetitive tasks at regular intervals.
```

Let's dive into a practical example within the context of a sales company.

> Imagine the company's goal is to send diverse sales reports to customers based on the specific reports each customer chooses to receive. In constructing the schedule for this scenario, the following steps are taken:

Reviewer: ```suggestion
Let's imagine the company's goal is to send diverse sales reports to customers based on the specific reports each customer chooses to receive. In constructing the schedule for this scenario, the following steps are taken:
```

#. Iterate over reports stored in the database and create a recurring task for each report, considering its unique properties. This task, however, should not be generated during holiday periods.

> #. Furthermore, we encounter another critical task that needs scheduling: the periodic cleanup of outdated files that are no longer relevant.

Reviewer: ```suggestion
#. The periodic cleanup of outdated files that are no longer relevant.
```

I think it's simpler yet understandable. This makes the sentence lighter :slightly_smiling_face:

On the basis of a case study in the context of a full stack Symfony application, let's dive in and explore how we can set up our system.

> Symfony Scheduler basics:

Reviewer: ```suggestion
The Scheduler Basics
```

-------------------------

> Differences and parallels between Symfony Messenger and Symfony Scheduler.

Reviewer: ```suggestion
Differences and Similarities Between Messenger and Scheduler
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The primary goal is to generate and process reports generation while also handling the removal of outdated reports at specified intervals.

> As mentioned, this component, even if it's an independent component, it draws its foundation and inspiration from the Symfony Messenger component.

Reviewer: I'm not sure this sentence brings much added value with the one that's following

> On one hand, it adopts well-established concepts from Symfony Messenger (such as message, handler, bus, transport, etc.). For example, the task of creating a report is considered as a message that will be directed, and processed by the corresponding handler.

Reviewer: ```suggestion
It adopts well-established concepts from Messenger (such as message, handler, bus, transport, etc.). For example, the task of creating a report is considered as a message that will be directed, and processed by the corresponding handler.
```

You can just speak about "Messenger" and remove "Symfony". Makes it lighter :slightly_smiling_face:

However, unlike Symfony Messenger, the messages will not be dispatched in the first instance. Instead, the aim is to create them based on a predefined frequency.

This is where the specific transport in Symfony Scheduler, known as the :class:`Symfony\\Component\\Scheduler\\Messenger\\SchedulerTransport`, plays a crucial role. The transport autonomously generates directly various messages according to the assigned frequencies.

From (Symfony Messenger cycle):

.. image:: /_images/components/messenger/basic_cycle.png
    :alt: Symfony Messenger basic cycle

To (Symfony Scheduler cycle):

.. image:: /_images/components/scheduler/scheduler_cycle.png
    :alt: Symfony Scheduler basic cycle

In essence, it is crucial to precisely define the message to be conveyed and processed. In Symfony Scheduler, the concept of a message takes on a very particular characteristic; it should be recurrent: It's a :class:`Symfony\\Component\\Scheduler\\RecurringMessage`.

> Clarifying the need to define and attach Recurring Messages to a Schedule.

Reviewer: ```suggestion
Attaching a Recurring Message To a Schedule
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In order to generate various messages based on their defined frequencies, configuration is necessary.
The heart of the scheduling process and its configuration resides in a class that must extend the :class:`Symfony\\Component\\Scheduler\\ScheduleProviderInterface`.

> The purpose of this provider is to return a schedule through the method ``getSchedule()`` containing our different recurringMessages.

Reviewer: You can use `:method:` here to automatically link the method to its definition on Github

> The class attribute ``AsSchedule``, which by default references the ``default`` named schedule, allows us to register on a particular schedule::

Reviewer: Same here for AsSchedule with `:class`:

```suggestion
The ``AsSchedule`` attribute, which by default references the ``default`` named schedule, allows us to register on a particular schedule::
```

    // src/Scheduler/MyScheduleProvider.php
    namespace App\Scheduler;

    class MyScheduleProvider implements ScheduleProviderInterface
    {
        public function getSchedule(): Schedule
        {

>             ...

Reviewer: ```suggestion
            // ...
```

        }
    }

.. tip::

    This becomes important when initiating the ``messenger:consume`` command, especially when specifying one or more specific transports.
    In Symfony Scheduler, the transport is named ``scheduler_nameofyourschedule``.

The Concept of Recurring Message in Symfony Scheduler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A message associated with a Trigger

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: I think this title can be removed

First and foremost, a RecurringMessage is a message that will be associated with a trigger.

The trigger is what allows configuring the recurrence frequency of your message. Several options are available to us:

> #. It can be a cron expression trigger:

Reviewer 2: We need to note that [dragonmantank/cron-expression](https://packagist.org/packages/dragonmantank/cron-expression) is required to use the cron expression trigger.

Might also be nice to have a link to https://crontab.guru/ to help users construct/understand cron expressions.

.. configuration-block::

    .. code-block:: php

>         RecurringMessage::cron(‘*****’, new Message())

Reviewer: ```suggestion
        RecurringMessage::cron('*****', new Message());
```

Reviewer 2: ```suggestion
        RecurringMessage::cron('* * * * *', new Message())
```

(spaces between are required)

        RecurringMessage::cron('#daily', $msg);

.. tip::

>     A hashed expressions is a string representing the schedule for a particular command to execute.

Reviewer 2: This part isn't quite correct. What you're describing is the standard cron aliases

- `@daily`
- `@monthly`
- etc...

Hashed aliases are different in that they create a random but deterministic standard cron expression based on some context (in this case, the message's `__toString()` value).

At some point later in the doc, we need to discuss the importance of not having many tasks all run at the same time (ie 10 jobs at 00:00 every day). The jitter trigger wrapper is one solution, another is hashed cron expressions. I think here is where we should describe this feature.

Why don't you drop this section for now (or just show more examples of _standard_ cron expressions) and I'll create a follow-up PR explaining hashed expressions. I originally ported this code to Symfony and use it quite extensively.

    Supported hashed expression:

    * ``#yearly``, ``#annually`` - Run once a year, midnight, Jan. 1 - ``0 0 1 1 *``
    * ``#monthly`` - Run once a month, midnight, first of month - ``0 0 1 * *``
    * ``#weekly`` - Run once a week, midnight on Sun - ``0 0 * * 0``
    * ``#daily``, ``#midnight`` - Run once a day, midnight - ``0 0 * * *``
    * ``#hourly`` - Run once an hour, first minute - ``0 * * * *``

.. tip::

>     It's possible to add and define a timezone as a 3rd argument

Reviewer: You may actually put this in the code snippet above, something like

```
// optionally specify the timezone to run the job on
RecurringMessage::cron...

Scheduler
=========

.. versionadded:: 6.3

    The Scheduler component was introduced in Symfony 6.3

> The Symfony Scheduler is a powerful and flexible component designed to manage tasks scheduling within your PHP application.

Reviewer 3: ```suggestion
Scheduler is a component designed to manage task scheduling within your PHP application - like running a task each night at 3 am, every 2 weeks except for holidays or any schedule you can imagine.
```

This document focuses on using the Scheduler component in the context of a full stack Symfony application.

Installation
------------

In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
install the scheduler component:

.. code-block:: terminal

    $ composer require symfony/scheduler

> Debugging the Schedule

Reviewer 3: Let's move this later. I haven't learned how to create recurring message yet.

    Additionally, you have the option to display terminated recurring messages using the ``--all`` option.

.. code-block:: terminal

    $ php bin/console debug:scheduler

      Scheduler
      =========

      default
      -------

        ------------------- ------------------------- ----------------------
        Trigger             Provider                  Next Run
        ------------------- ------------------------- ----------------------
        every 2 days        App\Messenger\Foo(0:17..)  Sun, 03 Dec 2023 ...
        15 4 */3 * *        App\Messenger\Foo(0:17..)  Mon, 18 Dec 2023 ...
       -------------------- -------------------------- ---------------------

Introduction to the case
------------------------

Embarking on a task is one thing, but often, the need to repeat that task looms large.
While one could resort to issuing commands and scheduling them with cron jobs, this approach involves external tools and additional configuration.

The Scheduler component emerges as a solution, allowing you to retain control, configuration, and maintenance of task scheduling within our PHP application.

> At its core, the principle is straightforward: a task, considered as a message needs to be managed by a service, and this cycle must be repeated.

Reviewer 3: The "considered as a message" doesn't make sense to me. Does schedule call them "messages" but the user thinks of them as tasks? If so, how about:

> At its core, scheduler allows you to create a task (called a message) that is executed by a service and repeated on some schedule.

Author: I understand your point of view. Indeed, what I meant to say is: 'At its core, the principle is straightforward: a task, treated as a message by the Scheduler, must be handled by a service, and this cycle should be repeated'

Does this sound familiar? Think :doc:`Symfony Messenger docs </components/messenger>`.

But while the system of Messenger proves very useful in various scenarios, there are instances where its capabilities

> fall short, particularly when dealing with repetitive tasks at regular intervals.

Reviewer 3: I think the writing - while very nice - is getting long in this section. I think (but perhaps I'm wrong) that the idea of scheduling tasks is fairly easy to understand. Pointing out that this could be done by a CRON but that scheduler allows you to keep your config all in one place & is more flexible is nice. But once we have that, we can get straight to using scheduler.

Let's dive into a practical example within the context of a sales company.

Imagine the company's goal is to send diverse sales reports to customers based on the specific reports each customer chooses to receive.
In constructing the schedule for this scenario, the following steps are taken:

#. Iterate over reports stored in the database and create a recurring task for each report, considering its unique properties. This task, however, should not be generated during holiday periods.

> #. Furthermore, you encounter another critical task that needs scheduling: the periodic cleanup of outdated files that are no longer relevant.

Reviewer 3: I like this idea of having a real-use case that is showing non-trivial items. But I think we need some basic usage docs up higher. It's a lot of reading if I'm trying to get a general idea of what this component does.

The primary goal is to generate and process reports generation while also handling the removal of outdated reports at specified intervals.

As mentioned, this component, even if it's an independent component, it draws its foundation and inspiration from the Messenger component.

On one hand, it adopts well-established concepts from Messenger (such as message, handler, bus, transport, etc.).
For example, the task of creating a report is considered as a message that will be directed, and processed by the corresponding handler.

However, unlike Messenger, the messages will not be dispatched in the first instance. Instead, the aim is to create them based on a predefined frequency.

This is where the specific transport in Scheduler, known as the :class:`Symfony\\Component\\Scheduler\\Messenger\\SchedulerTransport`, plays a crucial role.
The transport autonomously generates directly various messages according to the assigned frequencies.

From (Messenger cycle):

.. image:: /_images/components/messenger/basic_cycle.png
    :alt: Symfony Messenger basic cycle

To (Scheduler cycle):

.. image:: /_images/components/scheduler/scheduler_cycle.png

>     :alt: Symfony Scheduler basic cycle

Reviewer 3: I like these diagrams. But it may be too much for this early in the docs.

Author: Shouldn't we distinguish between the Messenger cycle and that of the Scheduler from the outset?

In essence, it is crucial to precisely define the message to be conveyed and processed. In Scheduler, the concept of a message takes on a very particular characteristic;
it should be recurrent: It's a :class:`Symfony\\Component\\Scheduler\\RecurringMessage`.

Clarifying the need to define and attach RecurringMessages to a Schedule
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In order to generate various messages based on their defined frequencies, configuration is necessary.
The heart of the scheduling process and its configuration resides in a class that must extend the :class:`Symfony\\Component\\Scheduler\\ScheduleProviderInterface`.

> The purpose of this provider is to return a schedule through the method :method:`Symfony\\Component\\Scheduler\\ScheduleProviderInterface::getMethod` containing your different recurringMessages.

Reviewer 3: `getSchedule()` method, right?

Author: 🙈😅

> The :class:`Symfony\\Component\\Scheduler\\Attribute\\AsSchedule` attribute, which by default references the ``default`` named schedule, allows you to register on a particular schedule::

Reviewer 3: Should we show the attribute below? Or is that optional?

Author: oh yes indeed, it's just an oversight. If you don't specify the name (just AsSchedule attribute), it will default to 'default'. The way it's written in the tip section can be misleading: I update it

    // src/Scheduler/MyScheduleProvider.php
    namespace App\Scheduler;

>     class MyScheduleProvider implements ScheduleProviderInterface

Reviewer 3: If we can think of a more realistic name, that would be awesome :)

    {
        public function getSchedule(): Schedule
        {
            // ...
        }
    }

.. tip::

    This becomes important when initiating the ``messenger:consume`` command, especially when specifying one or more specific transports.
    In Scheduler, the transport is named ``scheduler_nameofyourschedule``.

.. tip::

>     It is a good practice to memoize your schedule to prevent unnecessary reconstruction if the ``getSchedule`` method is checked by another service or internally within Symfony

Reviewer 3: I'm not sure we should mention this. Memoizing is a pretty technical term and probably creating a schedule is not a heavy thing anyway.

> The Concept of RecurringMessage in Scheduler

Reviewer 3: Are there any other types of messages that you would schedule? Suggestion for the title:

> Scheduling Recurring Messages

And, I like this section. This is where we get into the meat of things

                return null;
            }

            while (!$this->isHoliday($nextRun) { // loop until you get the next run date that is not a holiday
                $nextRun = $this->inner->getNextRunDate($nextRun);
            }

            return $nextRun;
        }

        private function isHoliday(\DateTimeImmutable $timestamp): bool
        {
            // app specific logic to determine if $timestamp is on a holiday
            // returns true if holiday, false otherwise
        }
    }

Then, you would have to define your RecurringMessage

.. configuration-block::

    .. code-block:: php

        RecurringMessage::trigger(
            new ExcludeHolidaysTrigger( // your custom trigger wrapper
                CronExpressionTrigger::fromSpec('@daily'),
            ),
            new SendDailySalesReports(),
        );

The RecurringMessages must be attached to a Schedule::

    // src/Scheduler/MyScheduleProvider.php
    namespace App\Scheduler;

    #[AsSchedule('uptoyou')]
    class MyScheduleProvider implements ScheduleProviderInterface
    {
        public function getSchedule(): Schedule
        {

>             return $this->schedule ??= (new Schedule())

Reviewer 3: Is the `$this->schedule` part of the memoization? I really need to be convinced that it is VERY important to keep this detail

                ->with(
                    RecurringMessage::trigger(
                        new ExcludeHolidaysTrigger( // your custom trigger wrapper
                            CronExpressionTrigger::fromSpec('@daily'),
                        ),
                    new SendDailySalesReports()),
                    RecurringMessage::cron(‘3 8 * * 1’, new CleanUpOldSalesReport())

                );
        }
    }

So, this RecurringMessage will encompass both the trigger, defining the generation frequency of the message, and the message itself, the one to be processed by a specific handler.

Efficient management with Symfony Scheduler
-------------------------------------------

> However, if your worker becomes idle, since the messages from your schedule are generated on-the-fly by the schedulerTransport,

Reviewer 3: We haven't talked about workers or `messenger:consume` yet. I think we should at least show them. The story I want to see is:

A) Creating a message & handler of some sort (this can be very quick and point to Messenger, and the handler can be empty, but we should show it)
B) Creating a schedule and setting up this message in a recurring way
C) Running `messenger:consume` - and talking a bit about how that works and what to expect.

    #[AsSchedule]
    class SaleTaskProvider implements ScheduleProviderInterface
    {
        public function getSchedule(): Schedule
        {
            // ...
        }
    }

.. tip::

    By default, if not specified, the schedule name will be ``default``.
    In Scheduler, the name of the transport is formed as follows: ``scheduler_nameofyourschedule``.

.. tip::

    It is a good practice to memoize your schedule to prevent unnecessary reconstruction if the ``getSchedule`` method is checked by another service or internally within Symfony

Scheduling Recurring Messages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

First and foremost, a RecurringMessage is a message that will be associated with a trigger.

The trigger is what allows configuring the recurrence frequency of your message. Several options are available to us:

#. It can be a cron expression trigger:

.. configuration-block::

    .. code-block:: php

        RecurringMessage::cron(‘* * * * *’, new Message());

.. tip::

    `dragonmantank/cron-expression`_ is required to use the cron expression trigger.

    Also, `crontab_helper`_ is a good tool if you need help to construct/understand cron expressions

> .. versionadded:: 6.4
> 
>     Since version 6.4, it is now possible to add and define a timezone as a 3rd argument

Reviewer 4: ```suggestion
```

To keep for 6.4 PR

Author: https://github.com/symfony/symfony-docs/pull/19244#issuecomment-1851974661 
It'll be updated in https://github.com/symfony/symfony-docs/pull/19292 + version added in 6.3

        );

The RecurringMessages must be attached to a Schedule::

    // src/Scheduler/MyScheduleProvider.php
    namespace App\Scheduler;

    #[AsSchedule('uptoyou')]
    class SaleTaskProvider implements ScheduleProviderInterface
    {
        public function getSchedule(): Schedule
        {
            return $this->schedule ??= (new Schedule())
                ->with(
                    RecurringMessage::trigger(
                        new ExcludeHolidaysTrigger( // your custom trigger wrapper
                            CronExpressionTrigger::fromSpec('@daily'),
                        ),
                    new SendDailySalesReports()),
                    RecurringMessage::cron(‘3 8 * * 1’, new CleanUpOldSalesReport())

                );
        }
    }

So, this RecurringMessage will encompass both the trigger, defining the generation frequency of the message, and the message itself, the one to be processed by a specific handler.

Debugging the Schedule
~~~~~~~~~~~~~~~~~~~~~~

The ``debug:scheduler`` command provides a list of schedules along with their recurring messages.
You can narrow down the list to a specific schedule.

> .. versionadded:: 6.4
> 
>     Since version 6.4, you can even specify a date to determine the next run date using the ``--date`` option.
>     Additionally, you have the option to display terminated recurring messages using the ``--all`` option.

Reviewer 4: ```suggestion
```

Scheduler
=========

.. versionadded:: 6.3

    The Scheduler component was introduced in Symfony 6.3

Scheduler is a component designed to manage task scheduling within your PHP application - like running a task each night at 3 am, every 2 weeks except for holidays or any schedule you can imagine.

This component proves highly beneficial for tasks such as database cleanup, automated maintenance (e.g., cache clearing), background processing (queue handling, data synchronization), periodic data updates, or even scheduled notifications (emails, alerts), and more.

This document focuses on using the Scheduler component in the context of a full stack Symfony application.

Installation
------------

In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
install the scheduler component:

.. code-block:: terminal

    $ composer require symfony/scheduler

Introduction to the case
------------------------

Embarking on a task is one thing, but often, the need to repeat that task looms large.

> While one could resort to issuing commands and scheduling them with cron jobs, this approach involves external tools and additional configuration.

Reviewer 5: I think the main benefit is that everything is managed in your app and that gives a lot of flexibility that is just not possible with cron jobs (dynamic schedules being the main one).

Author: yes, indeed, I should emphasize this point more. I'll update

> The Scheduler component emerges as a solution, allowing you to retain control, configuration, and maintenance of task scheduling within our PHP application.

Reviewer 5: ```suggestion
The Scheduler component emerges as a solution, allowing you to retain control, configuration, and maintenance of task scheduling within PHP applications.
```

> At its core, scheduler allows you to create a task (called a message) that is executed by a service and repeated on some schedule.

Reviewer 5: ```suggestion
At its core, the Scheduler component allows you to create a task (called a message) that is executed by a service and repeated on some schedule.
```

Scheduler
=========

.. versionadded:: 6.3

    The Scheduler component was introduced in Symfony 6.3

The Symfony Scheduler is a powerful and flexible component designed to manage tasks scheduling within your PHP application.

This document focuses on using the Scheduler component in the context of a full stack Symfony application.

Installation
------------

In applications using :ref:`Symfony Flex <symfony-flex>`, run this command to
install the scheduler component:

.. code-block:: terminal

    $ composer require symfony/scheduler

Debugging the Schedule
~~~~~~~~~~~~~~~~~~~~~~

The ``debug:scheduler`` command provides a list of schedules along with their recurring messages.

You can narrow down the list to a specific schedule, or even specify a date to determine the next run date using the ``--date`` option.

> Additionally, you have the option to display terminated recurring messages using the ``--all`` option.

Reviewer 4: The `--all` option has been added in 6.4. Maybe we should keep it for another PR ?

Reviewer 4: Seems like `--date` has been added at the same time, 6.4 too

        {
            // ...
        }
    }

.. tip::

    This becomes important when initiating the ``messenger:consume`` command, especially when specifying one or more specific transports.
    In Scheduler, the transport is named ``scheduler_nameofyourschedule``.

.. tip::

    It is a good practice to memoize your schedule to prevent unnecessary reconstruction if the ``getSchedule`` method is checked by another service or internally within Symfony

The Concept of RecurringMessage in Scheduler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A message associated with a Trigger
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

First and foremost, a RecurringMessage is a message that will be associated with a trigger.

The trigger is what allows configuring the recurrence frequency of your message. Several options are available to us:

#. It can be a cron expression trigger:

.. configuration-block::

    .. code-block:: php

        RecurringMessage::cron(‘* * * * *’, new Message());

.. tip::

    `dragonmantank/cron-expression`_ is required to use the cron expression trigger.

    Also, `crontab_helper`_ is a good tool if you need help to construct/understand cron expressions

.. tip::

>     It's possible to add and define a timezone as a 3rd argument

Reviewer 4: IIUC, timezone for cron expression trigger is a 6.4 feature

    class MyScheduleProvider implements ScheduleProviderInterface
    {
        public function getSchedule(): Schedule
        {
            return $this->schedule ??= (new Schedule())
                ->with(
                    RecurringMessage::trigger(
                        new ExcludeHolidaysTrigger( // your custom trigger wrapper
                            CronExpressionTrigger::fromSpec('@daily'),
                        ),
                    // instead of being static as in the previous example
                    new CallbackMessageProvider([$this, 'generateReports'], 'foo')),
                    RecurringMessage::cron(‘3 8 * * 1’, new CleanUpOldSalesReport())

                );
        }

        public function generateReports(MessageContext $context)
        {
            // ...
            yield new ReportSomething();
            yield new ReportSomethingReportSomethingElse();
            ....
        }
    }

Exploring alternatives for crafting your Recurring Messages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There is also another way to build a RecurringMessage, and this can be done simply by adding an attribute above a service or a command:

> :class:`Symfony\\Component\\Scheduler\\Attribute\\AsPeriodicTask` attribute and :class:`Symfony\\Component\\Scheduler\\Attribute\\AsCronTask` attribute.

Reviewer 4: Those attributes have been added in 6.4

Managing Scheduled Messages
---------------------------

Modifying Scheduled Messages in real time
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

While planning a schedule in advance is beneficial, it is rare for a schedule to remain static over time.
After a certain period, some RecurringMessages may become obsolete, while others may need to be integrated into our planning.

As a general practice, to alleviate a heavy workload, the RecurringMessages in the schedules are stored in memory to avoid recalculation each time the scheduler transport generates messages.
However, this approach can have a flip side.

In the context of our sales company, certain promotions may occur during specific periods and need to be communicated repetitively throughout a given timeframe
or the deletion of old reports needs to be halted under certain circumstances.

This is why the Scheduler incorporates a mechanism to dynamically modify the schedule and consider all changes in real-time.

Exploring Strategies for adding, removing, and modifying entries within the Schedule
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The schedule provides you with the ability to :method:`Symfony\\Component\\Scheduler\Schedule::add`, :method:`Symfony\\Component\\Scheduler\Schedule::remove`, or :method:`Symfony\\Component\\Scheduler\Schedule::clear` all associated RecurringMessages,

Reviewer 4: Methods `remove` and `clear` have been added in 6.4

        {
            // do what you have to do

            if ($isFinished) {
                $this->mySchedule->removeCleanUpMessage();
            }
        }
    }

Nevertheless, this system may not be the most suitable for all scenarios. Also, the handler should ideally be designed to process the type of message it is intended for,
without making decisions about adding or removing a new RecurringMessage.

For instance, if, due to an external event, there is a need to add a recurrent message aimed at deleting reports,
it can be challenging to achieve within the handler. This is because the handler will no longer be called or executed once there are no more messages of that type.

However, the Scheduler also features an event system that is integrated into a Symfony full-stack application by grafting onto Symfony Messenger events.
These events are dispatched through a listener, providing a convenient means to respond.

Managing Scheduled Messages via Events
--------------------------------------

A strategic event handling
~~~~~~~~~~~~~~~~~~~~~~~~~~

The goal is to provide flexibility in deciding when to take action while preserving decoupling.

> Three primary event types have been introduced types

Reviewer 4: Events have been added in 6.4

Introduction to the case:
-------------------------

- Provide a case scenario of sending tasks as messages to different handlers.
- Reference Symfony Messenger for the non recurring approach :doc:`Messenger component docs </components/messenger>`.
- To emphasize that this doesn't cover every situation and that certain cases will require recurrence, we can provide a few examples.

On the basis of a case study in the context of a full stack Symfony application, we will cover the following topics:

Symfony Scheduler basics:
-------------------------

Exploring differences and parallels between Symfony Messenger and Symfony Scheduler.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
- Outline the key differences and similarities between Symfony Messenger and Symfony Scheduler.
- Compare the different cycles.
- Introduce the Scheduler Transport and its particularity: generate directly the different messages to handle

The Concept of Recurring Message in Symfony Scheduler
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A message associated with a Trigger
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> - The importance of setting the frequency for your messages: Introduce the concept of a Trigger in Symfony Scheduler encompassing all the different types.

Reviewer 2: When covering custom triggers, an interesting example could be a trigger wrapper that excludes holidays: Cron task that runs every weekday at midnight, except "holidays" (defined by some kind of lookup table).

```php
RecurringMessage::trigger(
    new ExcludeHolidaysTrigger( // our custom trigger wrapper
        CronExpressionTrigger::fromSpec('@daily'),
    ),
    new SendDailySalesReports(),
);
```

And something like the following for the custom trigger (untested):

```php
class ExcludeHolidaysTrigger implements TriggerInterface
{
    public function __construct(private TriggerInterface $inner)
    {
    }

    public function __toString(): string
    {
        return $this->inner.' (except holidays)';
    }

    public function getNextRunDate(\DateTimeImmutable $run): ?\DateTimeImmutable
    {
        if (!$nextRun = $this->inner->getNextRunDate($run)) {
            return null;
        }

        while (!$this->isHoliday($nextRun) { // loop until we get the next run date that is not a holiday
            $nextRun = $this->inner->getNextRunDate($nextRun);
        }

        return $nextRun;
    }

    private function isHoliday(\DateTimeImmutable $timestamp): bool
    {
        // app specific logic to determine if $timestamp is on a holiday
        // returns true if holiday, false otherwise
    }
}
```
